### PR TITLE
MGMTBUGSM-191: Race between agent and clusterDeployment reconciler causes day2 host installation to ignore requested hostname

### PR DESCRIPTION
**Pull request by ori-amizur**
In kube API, in agent controller, there is a function updateIfNeeded
that updates host record in DB.  There are severeal updates in this
function.  The first is to synchronize the approved flag with the spec.

When the clusterDeployment controller attempts to install day2 host, it
checks if it ready and if it is approved.  If the answer answer to both
is yes, then it starts installation.

There might be a situation that agent flag is set to true, but not all
host updates have been performed yet (hostname in our case).  In this
case the install-config will not reflect completely the host data after
update.

The solution is to move the update of te approved flag to be the last
operation in that function (after all the other updates).

## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [x] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Assignees

/cc @yevgeny-shnaidman 
/cc @flaper87 

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/controller/controllers/agent_controller.go [hunk 1]
```diff
@@ -1086,18 +1086,6 @@ func (r *AgentReconciler) updateIfNeeded(ctx context.Context, log logrus.FieldLo
 	spec := agent.Spec
 	var err error
 
-	if internalHost.Approved != spec.Approved {
-		err = r.Installer.UpdateHostApprovedInternal(ctx, internalHost.InfraEnvID.String(), agent.Name, spec.Approved)
```
Comment (filanov) [inline internal/controller/controllers/agent_controller.go:1090]: I'm suggesting to return an finish the reconcile after this call. 
This way the reconcile will be short and no races will be done no?

Comment (ori-amizur) [reply internal/controller/controllers/agent_controller.go:1090]: When should the host update be done ?

Comment (ori-amizur) [reply internal/controller/controllers/agent_controller.go:1090]: The fix should verify that the approved flag is set not before all other host updates

**Review by filanov — APPROVED** [review]

**Comment by filanov** [comment]
@ori-amizur please retitle the PR to bz so we can backport it later

**Comment by ori-amizur** [comment]
/retitle Bug 2064622: Race between agent and clusterDeployment reconciler causes day2 host installation to ignore requested hostname

**Comment by filanov** [comment]
@ori-amizur it was pushed to 2.5 so you don't need to backport

**Pull request merged**